### PR TITLE
Do not suggest where keyword in generic constraints when accessing members

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/30785

### CODE REVIEW

File: src/EditorFeatures/CSharpTest2/Recommendations/WhereKeywordRecommenderTests.cs [hunk 1]
```diff
@@ -297,6 +297,27 @@ public async Task TestNoWhereAfterDot()
...
+    void M<T> where T : System.$$
+}");
+        }
+
+        [WorkItem(30785, "https://github.com/dotnet/roslyn/issues/30785")]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        public async Task TestNotAfterDot2()
+        {
+            await VerifyAbsenceAsync(
+@"public class C<T> where T : System.$$
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhereKeywordRecommenderTests.cs:319]: consider test for `where T : global::` and `where T : List<$$`,  `where T : List<(int, $$` etc.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs [hunk 2]
```diff
@@ -109,13 +109,16 @@ private static bool IsTypeParameterConstraintContext(CSharpSyntaxContext context
...
             }
 
-            // class C<T> where T : IGoo |
-            // delegate void D<T> where T : IGoo |
-            var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();
-
-            if (constraintClause != null)
+            // Do not check constraints right after dot token, as it is probably member access.
+            // See https://github.com/dotnet/roslyn/issues/30785 for example
+            if (!token.IsKind(SyntaxKind.DotToken))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs:114]: i think what probably makes more sense is asking if we're after the end token of a *complete* constraint.

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
Let's try for a more generalized construct.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs [hunk 3]
```diff
@@ -109,13 +109,17 @@ private static bool IsTypeParameterConstraintContext(CSharpSyntaxContext context
...
 
-            // class C<T> where T : IGoo |
-            // delegate void D<T> where T : IGoo |
-            var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();
-
-            if (constraintClause != null)
+            // Check constraints only if target token is the last in the TypeParameterConstraintSyntax node.
+            // Otherwise we can be in incomplete constraint and "where" keyword will not be valid suggestion.
+            // See https://github.com/dotnet/roslyn/issues/30785 for example
+            if (token.IsLastTokenOfNode<TypeParameterConstraintSyntax>())
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs:115]:
```suggestion
            if (token.IsLastTokenOfNode<TypeParameterConstraintSyntax>(out var constraintClause))
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs [hunk 4]
```diff
@@ -109,13 +109,17 @@ private static bool IsTypeParameterConstraintContext(CSharpSyntaxContext context
...
             {
-                if (constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
+                // class C<T> where T : IGoo |
+                // delegate void D<T> where T : IGoo |
+                var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();
+
+                if (constraintClause is not null &&
+                    constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs:122]:
```suggestion
                var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();

                if (constraintClause is not null &&
                    constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
```

can't all of thise be dleeted now?

Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs]: Indeed!

**Review by CyrusNajmabadi — APPROVED** [review]
Awesome.  Thank you!

**Pull request merged**